src/common/stack.rs

```diff
@@ -1,88 +1,419 @@
+		self.stack.swap_remove(index)
+	}
+
+	pub(crate) fn top(&self) -> Option<&T> {
+		self.stack.last()
+	}
+
+	pub(crate) fn top_mut(&mut self) -> Option<&mut T> {
+		self.stack.last_mut()
+	}
+
+	// Same as Vec::[truncate](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate)
+	pub(crate) fn truncate(&mut self, new_size: usize) {
+		self.stack.truncate(new_size)
+	}
+
+	#[inline]
+	pub(crate) fn len(&self) -> usize {
+		self.stack.len()
 	}
-}
 
-#[cfg(feature = "std")]
-impl error::Error for Error {
-	fn description(&self) -> &str {
-		&self.0
+	pub(crate) fn is_empty(&self) -> bool {
+		self.stack.is_empty()
 	}
 }
 
-/// Stack with limit.
-#[derive(Debug)]
-pub struct StackWithLimit<T> where T: Clone {
-	/// Stack values.
-	values: Vec<T>,
-	/// Stack limit (maximal stack len).
-	limit: usize,
+// Why introduce the extra complexity of StackSizeLimit, StackSizeInitial, and StackSize?
+// We want to make the user to do the correct thing using the type checker.
+// Check out the excellent Rust API Guidelines (linked below) for suggestions
+// about type safety in rust.
+//
+// By introducing the new typed arguments, we turn:
+//
+// ```
+// pub fn new(initial_size: usize, limit: usize) -> StackWithLimit<T> { ... }
+// ```
+//
+// into:
+//
+// ```
+// pub fn new(initial_size: StackSizeInitial<T>, limit: StackSizeLimit<T>) -> StackWithLimit<T> { ... }
+// ```
+//
+// https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-custom-type
+
+/// Type for communicating the size of some contigous container.
+/// Used for constructing both [`StackSizeLimit`] and [`StackSizeInitial`].
+#[derive(Eq, PartialEq, Hash, Debug)]
+pub struct StackSize<T> {
+	num_elements: usize,
+	// PhantomData is a zero-sized type for keeping track of T without rustc complaining.
+	phantom: PhantomData<*const T>,
 }
 
-impl<T> StackWithLimit<T> where T: Clone {
-	pub fn with_limit(limit: usize) -> Self {
-		StackWithLimit {
-			values: Vec::new(),
-			limit: limit
+impl<T> Clone for StackSize<T> {
+	fn clone(&self) -> Self {
+		StackSize {
+			num_elements: self.num_elements,
+			phantom: PhantomData,
 		}
 	}
+}
 
-	pub fn is_empty(&self) -> bool {
-		self.values.is_empty()
+impl<T> Copy for StackSize<T> {}
+
+impl<T> StackSize<T> {
+	/// Create StackSize based on number of elements.
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::StackSize;
+	/// let ss = StackSize::<(u8, u8)>::from_element_count(10);
+	/// assert_eq!(ss.element_count(), 10);
+	/// ```
+	pub fn from_element_count(num_elements: usize) -> StackSize<T> {
+		StackSize {
+			num_elements,
+			phantom: PhantomData,
+		}
 	}
 
-	pub fn len(&self) -> usize {
-		self.values.len()
+	/// Compute StackSize based on allowable memory.
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::StackSize;
+	/// let ss = StackSize::<(u8, u8)>::from_byte_count(10);
+	/// assert_eq!(ss.element_count(), 10 / 2);
+	/// ```
+	///
+	/// # Errors
```

Comment:
This header should be "Panics" but kudos for actually adding this section (too many people forget).

Reply:
fixed

Reply:
Does not seem to be fixed